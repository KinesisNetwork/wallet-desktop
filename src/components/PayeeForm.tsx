import { kebabCase, startCase } from 'lodash'
import * as React from 'react'

import { InputField } from '@components/InputField'
import { InputError } from '@helpers/errors'
import { FormAlert, Payee, WalletView } from '@types'

export interface Props {
  payee: Payee
  addPayee: (payee: Payee) => any
  handleChange: (field: keyof Payee, newValue: string) => any
  activeWalletView: WalletView
  cancelForm: () => any
  formIsInvalid: (formAlertArgs: FormAlert) => any
}

export class PayeeForm extends React.Component<Props> {
  constructor(props: Props) {
    super(props)
  }

  createNewPayee = (ev: React.FormEvent<HTMLFormElement>) => {
    ev.preventDefault()
    try {
      this.validateProps()
      this.props.addPayee(this.props.payee)
    } catch (e) {
      this.props.formIsInvalid({
        message: e.message,
        key: e.key
      })
    }
  }

  validateProps = (): void | never => {
    this.checkValidEntry('name')
    this.checkValidEntry('publicKey')
  }

  checkValidEntry = (key: keyof Payee) => {
    if (!this.props.payee[key]) {
      throw new InputError(`${startCase(key)} is required`, `input-payee-${kebabCase(key)}`)
    }
  }

  renderCancelButton = () => {
    return (
      <div className='control'>
        <button className='button is-danger' type='button' onClick={this.props.cancelForm}>Back</button>
      </div >
    )
  }

  render() {
    const { payee, handleChange } = this.props
    return (
      <div className='is-centered'>
        <form onSubmit={this.createNewPayee}>
          <InputField
            label='Payee Name'
            value={payee.name}
            id='payee-name'
            helpText='Name your payee'
            onChangeHandler={(newValue) => handleChange('name', newValue)}
          />
          <InputField
            label='Payee Public Key'
            value={payee.publicKey}
            id='payee-public-key'
            helpText='Set the public key of your payee'
            onChangeHandler={(newValue) => handleChange('publicKey', newValue)}
          />
          <div className='field is-grouped'>
            <div className='control is-expanded'>
              <button className='button is-fullwidth' type='submit'>Set Payee</button>
            </div>
            {this.props.activeWalletView !== WalletView.payees && this.renderCancelButton()}
          </div>
        </form>
      </div>
    )
  }
}
