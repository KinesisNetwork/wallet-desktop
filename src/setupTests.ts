import { configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

export default configure({ adapter: new Adapter() })
