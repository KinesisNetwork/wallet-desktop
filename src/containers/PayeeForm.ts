import { connect } from 'react-redux'

import { addPayee, changeTransferView, invalidForm, updatePayeeForm } from '@actions'
import { PayeeForm as PayeeFormPresentation } from '@components/PayeeForm'
import { Dispatch, RootState } from '@store'
import { FormAlert, Payee, TransferView } from '@types'

const mapStateToProps = ({ payees, view }: RootState) => ({
  payee: payees.form,
  activeWalletView: view.walletView,
})

const mapDispatchToProps = (dispatch: Dispatch) => ({
  addPayee: (payee: Payee) => dispatch(addPayee(payee)),
  handleChange: (field: keyof Payee, newValue: string) =>
    dispatch(updatePayeeForm({ field, newValue })),
  cancelForm: () => dispatch(changeTransferView(TransferView.transfer)),
  formIsInvalid: (error: FormAlert) => dispatch(invalidForm(error)),
})

export const PayeeForm = connect(
  mapStateToProps,
  mapDispatchToProps,
)(PayeeFormPresentation)
