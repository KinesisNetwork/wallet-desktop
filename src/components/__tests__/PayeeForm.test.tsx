import { shallow } from 'enzyme'
import * as React from 'react'

import { PayeeForm } from '@components/PayeeForm'
import '../../setupTests'

describe('PayeeForm', () => {
  let props

  beforeEach(() => {
    props = {
      activeWalletView: 3,
      payee: {
        name: '',
        publicKey: '',
      },
      addPayee: () => null,
      cancelForm: () => null,
      handleChange: () => null,
      callFormAlert: () => null,
    }
  })

  it('renders correctly', () => {
    const wrapper = shallow(<PayeeForm {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('returns a form with two InputField components and a button', () => {
    const wrapper = shallow(<PayeeForm {...props} />)

    const form = wrapper.find('form')

    expect(form.children()).toHaveLength(3)
    expect(form.find('InputField')).toHaveLength(2)
    expect(form.find('button')).toHaveLength(1)
    expect(form.find('button').text()).toEqual('Set Payee')
  })

  it("assigns the values of payee to InputField component's value", () => {
    const wrapper = shallow(
      <PayeeForm
        {...props}
        payee={{
          name: 'Fred',
          publicKey: '123',
        }}
      />,
    )

    const inputFields = wrapper.find('InputField')

    expect(inputFields.at(0).prop('value')).toEqual('Fred')
    expect(inputFields.at(1).prop('value')).toEqual('123')
  })

  it('calls the createNewPayee method on submitting the form', () => {
    const newPayee = {
      name: 'Fred',
      publicKey: '123',
    }
    const createNewPayeeMock = jest.fn()
    const wrapper = shallow(<PayeeForm {...props} payee={newPayee} addPayee={createNewPayeeMock} />)

    const form = wrapper.find('form')

    form.simulate('submit', { preventDefault: () => null })
    expect(createNewPayeeMock).toHaveBeenCalledWith(newPayee)
  })

  it('calls formAlert if validation is unsuccessful', () => {
    const messageAndKey = {
      message: 'Name is required',
      key: 'input-payee-name',
    }
    const formAlertMock = jest.fn()
    const wrapper = shallow(<PayeeForm {...props} formIsInvalid={formAlertMock} />)

    const form = wrapper.find('form')

    form.simulate('submit', { preventDefault: () => null })
    expect(formAlertMock).toHaveBeenCalledWith(messageAndKey)
  })

  describe('renders a cancel button', () => {
    it('if active wallet view is not set to payees', () => {
      const wrapper = shallow(<PayeeForm {...props} activeWalletView={2} />)

      const buttons = wrapper.find('button')

      expect(buttons).toHaveLength(2)

      const cancelButton = buttons.at(1)

      expect(cancelButton.hasClass('is-danger')).toBe(true)
      expect(cancelButton.text()).toEqual('Back')
    })

    it('and calls cancelForm button on click', () => {
      const cancelFormMock = jest.fn()
      const wrapper = shallow(
        <PayeeForm {...props} activeWalletView={2} cancelForm={cancelFormMock} />,
      )

      const cancelButton = wrapper.find('.is-danger')
      cancelButton.simulate('click')

      expect(cancelFormMock).toHaveBeenCalled()
    })
  })
})
