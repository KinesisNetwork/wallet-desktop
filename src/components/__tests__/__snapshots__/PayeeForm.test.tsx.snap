// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayeeForm renders correctly 1`] = `
<div
  className="is-centered"
>
  <form
    onSubmit={[Function]}
  >
    <InputField
      helpText="Name your payee"
      id="payee-name"
      label="Payee Name"
      onChangeHandler={[Function]}
      value=""
    />
    <InputField
      helpText="Set the public key of your payee"
      id="payee-public-key"
      label="Payee Public Key"
      onChangeHandler={[Function]}
      value=""
    />
    <div
      className="field is-grouped"
    >
      <div
        className="control is-expanded"
      >
        <button
          className="button is-fullwidth"
          type="submit"
        >
          Set Payee
        </button>
      </div>
    </div>
  </form>
</div>
`;
